sequential/utils.py

```diff
@@ -243,3 +248,110 @@ def validate_max_span(max_span: Union[int, None]):
+                                                    "Current minimum frequency is {}.".format(min_frequency)))
+        check_true(min_frequency * num_rows >= 1.0, ValueError("Minimum frequency percentage should set the minimum "
+                                                               "row count to be no less than 1.0. "
+                                                               "Thus the percentage should be no less than "
+                                                               "1/(number of sequences)."))
+    elif isinstance(min_frequency, int):
+        check_true(0 < min_frequency, ValueError("Minimum frequency should be greater than 0.0. "
+                                                 "Current minimum frequency is {}.".format(min_frequency)))
+        check_true(min_frequency <= num_rows, ValueError("Minimum frequency cannot be more than number of sequences. "
+                                                         "Current minimum frequency is {}.".format(min_frequency)))
+    else:
+        raise TypeError("Minimum frequency should be integer (as a row count) or float (as a row percentage).")
+
+
+def validate_min_frequency_with_batch(num_rows, batch_size, min_frequency):
+    """
+    Validate min_frequency when Seq2Pat runs on batches
+
+    """
+    if not isinstance(min_frequency, float) and num_rows > batch_size:
+        raise TypeError("Minimum frequency should be float (as a row percentage) when Seq2Pat runs on batches.")
+
+    try:
+        validate_min_frequency(batch_size, min_frequency)
+    except ValueError:
+        raise ValueError("Minimum frequency {} is not validate for the sequences "
+                         "in the chunk size {}!".format(min_frequency, batch_size))
+
+    remain_chunk_size = num_rows % batch_size
+    if remain_chunk_size > 1:
+        try:
+            validate_min_frequency(remain_chunk_size, min_frequency)
+        except ValueError:
+            raise ValueError("Minimum frequency {} is not validate for the sequences "
+                             "in the chunk size {}!".format(min_frequency, remain_chunk_size))
+
+
+def update_min_frequency(num_rows, min_frequency, min_frequency_df):
+    """
+    Update min_frequency when Seq2Pat runs on batches.
+
+    """
+    if isinstance(min_frequency, float):
+        min_frequency = max(min_frequency * min_frequency_df, 1.0/num_rows)
+    return min_frequency
+
+
+def list_to_counter(patterns):
+    """
+    Transform patterns to counter
+    """
+    patterns_without_frequency = list(map(lambda x: tuple(x[:-1]), patterns))
+    frequencies = list(map(lambda x: x[-1], patterns))
+
+    return collections.Counter(dict(zip(patterns_without_frequency, frequencies)))
+
+
+def counter_to_list(patterns_counter, min_row_count):
+    """
+    Transform counter of patterns to list
+    """
+    results = []
+    for key, value in patterns_counter.items():
```

Comment:
same q as before

Reply:
I missed this question before. So I tried an implementation of vectorizing, e.g. transform the dictionary to an array and then do the array operations instead of for-loop, but it seems the two methods have similar performance in tests, and the for-loop is slightly faster. Maybe it is because I still need to transform back to list of list as the final format in returns, where I need a loop. So I might keep this implementation for now until we find a better solution.

Reply:
yep, I see